aten/src/ATen/native/TensorCompare.cpp

```diff
@@ -243,15 +243,15 @@ std::vector<Tensor> where(const Tensor& condition) {
 }
 
 Tensor _s_where(const Tensor& condition, const Tensor& self, const Tensor& other) {
-  TORCH_CHECK(self.dtype() == other.dtype(), "expected scalar type ", self.dtype(), " but found ", other.dtype());
   Tensor ret = at::empty(self.sizes(), self.options());
   auto iter = at::TensorIteratorConfig()
-    .check_all_same_dtype(false)
     .set_check_mem_overlap(true)
     .add_output(ret)
     .add_input(condition)
     .add_input(self)
     .add_input(other)
+    .promote_inputs_to_common_dtype(true)
+    .cast_common_dtype_to_outputs(true)
```

Comment:
@mruberry 

I thought the change to support type-promotion will be as trivial as setting these flags for the `TensorIterator`. It seems to work fine with CUDA, however with CPU the error below is raised (error looks similar to one mentioned in #42660). Would be great if you can tell me why that assert is needed and also is there any preferred way to mitigate that. 

Thanks!!!
```
RuntimeError: !needs_dynamic_casting<func_t>::check(iter) INTERNAL ASSERT FAILED at "..\\aten\\src\\ATen/native/cpu/Loops.h":194, please report a bug to PyTorch. 
Exception raised from cpu_kernel at ..\aten\src\ATen/native/cpu/Loops.h:194 (most recent call first):
```

Code Reference of the Failed Assert
https://github.com/pytorch/pytorch/blob/f4695203c2c38864753873802a0835cfb7bc3dde/aten/src/ATen/native/cpu/Loops.h#L194

https://github.com/pytorch/pytorch/blob/f4695203c2c38864753873802a0835cfb7bc3dde/aten/src/ATen/native/TensorIteratorDynamicCasting.h#L15-L59

Reply:
Is it because you're calling the where_kernel with the condition's scalar type and not the iter's dtype?

Reply:
Maybe I can try that. Thank You for the quick response!

Reply:
Sadly it didn't work.
```python
>>> import torch
>>> a = torch.randn(3)
>>> a
tensor([ 0.4270, -0.3419, -0.4943])
>>> z = torch.zeros(3)
>>> z
tensor([0., 0., 0.])
>>> a.dtype
torch.float32
>>> torch.where(a > 0, a.to(torch.double), z)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: !needs_dynamic_casting<func_t>::check(iter) INTERNAL ASSERT FAILED at "../aten/src/ATen/native/cpu/Loops.h":194, please report a bug to PyTorch. 
>>> a = a.cuda()
>>> z = z.cuda()
>>> torch.where(a > 0, a.to(torch.double), z)
tensor([0.4270, 0.0000, 0.0000], device='cuda:0', dtype=torch.float64)
```

Reply:
Do we need to have some logic to manually take care of it for CPU?

Reduction Op seem to do something like that :thinking: 
https://github.com/pytorch/pytorch/blob/776c2d495f11ada421788b103043dd7ea11d4f39/aten/src/ATen/native/ReduceOpsUtils.h#L245

Reply:
Huh. So this check is necessary. CPU kernels require that type promotion happen first because they can't fuse the casts. CUDA kernels, on the other hand, can accept input in any dtype and then cast it in the kernel to the appropriate type.

I'm not totally sure what's going on here. I would try printing the dtypes of the tensors before they reach the kernel to see if they're still different and, if they are, why.

Reply:
Thank you for the suggestion, 
I checked that in where CPU kernel by adding the following. 
```cpp
std::cout << iter.input_dtype(0) << "\n";
std::cout << iter.input_dtype(1) << "\n";
std::cout << iter.input_dtype(2) << "\n";
std::cout << iter.output(0).dtype() << "\n";
```

```
Double
Double
Double
double
```

I think what is going on here is that the `condition` is also participating in the promotion.
So I was wondering is there a way to disable a particular input from taking part in the type promotion logic?
I couldn't see anything relevant from a quick glance at the Iterator code.

```cpp
Tensor _s_where(const Tensor& condition, const Tensor& self, const Tensor& other) {
  Tensor ret = at::empty(self.sizes(), self.options());
  auto iter = at::TensorIteratorConfig()
    .set_check_mem_overlap(true)
    .add_output(ret)
    .add_input(condition)
    .add_input(self)
    .add_input(other)
    .promote_inputs_to_common_dtype(true)
    .cast_common_dtype_to_outputs(true)
    .build();
  where_kernel(iter.device_type(), iter, iter.dtype()); // Note : iter.dtype()
  return ret;
```
Thank You!

Reply:
To answer your question and add some more detail, the system works like this:

- type promotion is either on or off (no partial type promotion supported currently)
- if type promotion is on, compute a "common dtype" for all inputs
- if on the CPU, create a copy of each input tensor in the "common dtype"
- pass the tensors to the CPU and CUDA implementations
- if on CPU, maybe cast the result to the desired output dtype

On CUDA we actually fuse the input and output tensor casts. 

To support type promotion for only some inputs you'd have to extend the TensorIterator interface. I'm not sure that's worthwhile to pursue at this time.

Reply:
Thank you for the detailed explanation. That is really helpful.

> To support type promotion for only some inputs you'd have to extend the TensorIterator interface. I'm not sure that's worthwhile to pursue at this time.

I have actually done that. Have updated the signature of these two methods.
```
TensorIteratorConfig& add_output(const Tensor& output, bool participate_promotion = false);
TensorIteratorConfig& add_input(const Tensor& input, bool participate_promotion = true);
```

And laid some simple machinery for not using a particular input in type-promotion.

This does seem like a big change as TensorIterator is very heavily used. However I think it would be more complicated to correctly support type-promotion for `where` without using TensorIterator functionality.

Let me know what you feel is the best way to go forward.

Thank You!!

Reply:
Additional detail: out arguments don't participate in type promotion, because the output shouldn't affect the computation.

Reply:
Right. That is why I had made it default to false. I see, even if we set it to `true`, it won't really do anything. Will remove that argument. Thanks!

Reply:
Hey @kshitij12345, I had a chance to talk to a few other engineers and we're very leery of changing how TensorIterator's type promotion works to handle this case. It's been a challenging area to write code correctly in and it requires extreme performance scrutiny. I know you've been thinking about this change and are pursuing a reasonable approach, but we'd like to encourage you to consider other challenges/tasks/issues that are better supported by PyTorch's current architecture and probably higher priority, too. I'm always happy to help you find something that would interest you, too.

Reply:
Sure. I understand. `TensorIterator` is hotspot interface for many Ops and I was anyways expecting rigorous testing and benchmarking for that change. 

However, one thing I'd like to note is that going forward, if Ops are getting numpy-like `where` argument support.
Eg. https://numpy.org/doc/stable/reference/generated/numpy.sin.html. This mechanism could be useful or maybe theres a better way :thinking: .

> I'm always happy to help you find something that would interest you, too.

Thanks! Would be great. Even I'll try to find something.

Reply:
I agree it's an interesting approach, and possibly one we'll have to adopt eventually.

If you can tell me a little about what you're looking for I could make some suggestions. for fun things to do.